File: src-lang/xt/ui/frames/table.clj
(ns xt.ui.frames.table
  "Schema-driven operational table frame."
  (:require [hara.lang :as l]))

(l/script :xtalk
  {:require [[xt.lang.spec-base :as xt]
             [xt.ui.core :as ui]]})

(defn.xt view [frame state actions]
  (var collection (or (. state ["collection"]) state))
  (var columns (or (. frame ["opts"] ["columns"]) []))
  (var header [])
  (xt/for:array [column columns]
    (xt/x:arr-push header
                   (ui/node "ui/table-cell"
                            {"value" (or (. column ["label"])
                                         (. column ["id"]))} [])))
  (var rows [])
  (xt/for:array [item (or (. collection ["items"]) [])]
    (var cells [])
    (xt/for:array [column columns]
      (xt/x:arr-push cells
                     (ui/node "ui/table-cell"
                              {"value" (xt/x:get-key item (. column ["id"]))}
                              [])))
    (xt/x:arr-push rows
                   (ui/node "ui/table-row"
                            {"key" (. item ["id"])}
                            cells)))
  (return
   (ui/node "ui/column" {"class" "gap-4"}
            [(ui/slot (xt/x:cat (. frame ["id"]) "/toolbar") [] {})
             (ui/node "ui/table" {"class" "w-full"}
                      [(ui/node "ui/table-header" {} header)
                       (ui/node "ui/table-body" {} rows)])])))

File: packages-gen/dart/libs/xt-ui/lib/frames/table.dart
import 'package:xtalk_ui/core.dart' as ui;


view(frame, state, actions) {
  var collection = state["collection"] ?? state;
  var columns = (frame["opts"])["columns"] ?? <dynamic>[];
  var header = <dynamic>[];
  var arr_52878 = columns;
  for(var i52879 = 0; i52879 < arr_52878.length; ++i52879){
    var column = arr_52878[i52879];
    header.add(ui.node(
      "ui/table-cell",
      <dynamic, dynamic>{"value":column["label"] ?? column["id"]},
      <dynamic>[]
    ));
  };
  var rows = <dynamic>[];
  var arr_52900 = collection["items"] ?? <dynamic>[];
  for(var i52901 = 0; i52901 < arr_52900.length; ++i52901){
    var item = arr_52900[i52901];
    var cells = <dynamic>[];
    var arr_52922 = columns;
    for(var i52923 = 0; i52923 < arr_52922.length; ++i52923){
      var column = arr_52922[i52923];
      cells.add(ui.node(
        "ui/table-cell",
        <dynamic, dynamic>{"value":item[column["id"]]},
        <dynamic>[]
      ));
    };
    rows.add(
      ui.node("ui/table-row",<dynamic, dynamic>{"key":item["id"]},cells)
    );
  };
  return ui.node("ui/column",<dynamic, dynamic>{"class":"gap-4"},<dynamic>[
    ui.slot(frame["id"] + "/toolbar",<dynamic>[],<dynamic, dynamic>{}),
    ui.node("ui/table",<dynamic, dynamic>{"class":"w-full"},<dynamic>[
      ui.node("ui/table-header",<dynamic, dynamic>{},header),
      ui.node("ui/table-body",<dynamic, dynamic>{},rows)
    ])
  ]);
}